"git merge-base"-based native build skipping

rye opened the pull request:
This PR changes the logic behind native build skipping to involve using `git merge-base` to find the latest common point in history to use as the beginning of a range log, rather than comparing strings.

This PR also removes the explicit halt if the current branch is called `master`, instead using a general purpose algorithm that should work in all cases, assuming the history is unshallowed enough to manage finding the beginning of a branch. (If a clone is shallow to depth `n`, a branch can be no longer than `n - 1`.)

This plumbing command is used internally for finding the best ancestor to start a merge at, so it makes a _lot_ of sense for us to use that when trying to figure out where we should start our comparison.

rye committed:
should_build?: Don't explicitly compare branches

This is a somewhat strange way of doing things.  We should be smarter.

Signed-off-by: Kristofer Rye <kristofer.rye@gmail.com>

rye committed:
should_build?: Add a DEFAULT_BRANCH constant

To tune at our leisure.

Signed-off-by: Kristofer Rye <kristofer.rye@gmail.com>

rye committed:
should_build?: Stop doing branch name comparison

Instead, use git merge-base to find the oldest shared point in history
and then use that as the basis of comparison.  Also, compare with
`HEAD` instead of some branch name.

[GitHub] This commit resolves #3253.

Signed-off-by: Kristofer Rye <kristofer.rye@gmail.com>

rye committed:
should_build: Remove now-unused git_branch method

Signed-off-by: Kristofer Rye <kristofer.rye@gmail.com>

rye committed:
should_build?: Refactor language in comments and skippable case

Signed-off-by: Kristofer Rye <kristofer.rye@gmail.com>

rye committed:
should_build?: Remove redundant return

Apparently RuboCop dislikes that style more.

Signed-off-by: Kristofer Rye <kristofer.rye@gmail.com>

rye committed:
should_build?: Add some commentary

Signed-off-by: Kristofer Rye <kristofer.rye@gmail.com>

rye committed:
Rename should-build to should_build

At RuboCop's suggestion.  Filenames should be snake_cased.  The method
is also named this, so it makes sense for them to be the same.

Signed-off-by: Kristofer Rye <kristofer.rye@gmail.com>

hawkrives commented on .circleci/config.yml:
```diff
@@ -343,7 +343,7 @@ jobs:
       - run: 
           name: Skip build if possible
-          command: ruby scripts/should-build.rb || circleci step halt
+          command: ruby scripts/should_build.rb || circleci step halt
```
it annoys me when cli tools use underscores. i get that rubocop complains the other way, but it annoys me.

rye replied:
Yes, but this is not just a cli tool, it's also a library file.

If we were instead to call it `scripts/should-build` with no file extension (which we _can_ do if it's `+x` and includes the `#!` line it needs) then RuboCop would not complain. I want to do that, if that sounds good to you.

hawkrives replied:
I would prefer that. 

Actually, if you wanted (probably in another PR), we could split this into a library file and an executable script – `should_build.rb` and `should-build`, and then `should-build` can require and invoke the function from `should_build.rb`.

rye replied:
That's a better way to do it anyhow, because then we can mock `sh` and such functions and test our actual behavior!

I'll rename the script as the last commit in this PR, then merge when green.

hawkrives replied:
(speaking of merge-when-green… #3242)

rye committed:
Rename should_build script to should-build, make executable

Signed-off-by: Kristofer Rye <kristofer.rye@gmail.com>

rye committed:
circle: Update to match changes to script names

Signed-off-by: Kristofer Rye <kristofer.rye@gmail.com>

rye committed:
CHANGELOG: Add a note about the changes in this PR

Signed-off-by: Kristofer Rye <kristofer.rye@gmail.com>

rye committed:
Negate should-<>-build logic

That is:

- Rename scripts/should-build to scripts/should-skip-build

- Flip logic in scripts/should-skip-build to exit with code 0 if the
  script should be skipped, and exit with code 1 if the script should
  not.

  - If the script fails to run for some reason, the default behavior
    is now to *not* skip the build.

Signed-off-by: Kristofer Rye <kristofer.rye@gmail.com>
